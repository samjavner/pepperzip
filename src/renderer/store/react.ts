import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
