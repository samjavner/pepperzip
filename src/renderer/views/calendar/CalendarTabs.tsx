import { Button, ButtonGroup } from "@blueprintjs/core";
import classNames from "classnames";
import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";
import CalendarPage from "./calendar/CalendarPage";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeCalendarPage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onCalendarClick: () =>
            dispatch(actions.setActiveCalendarPage("calendar")),
    };
}

type CalendarTabsProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const CalendarTabs: React.SFC<CalendarTabsProps> = props => (
    <>
        <div className="pz-navtabs">
            <ButtonGroup minimal={true}>
                <Button
                    className={classNames({
                        "pz-navtabs-button": true,
                        "pz-navtabs-button--active":
                            props.activePage === "calendar",
                    })}
                    icon="calendar"
                    onClick={props.onCalendarClick}
                >
                    Calendar
                </Button>
            </ButtonGroup>
        </div>
        <div style={{ borderTop: "1px solid lightgray" }} />
    </>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CalendarTabs);
