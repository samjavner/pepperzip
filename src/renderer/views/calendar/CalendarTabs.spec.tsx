import * as React from "react";
import * as renderer from "react-test-renderer";
import { CalendarTabs } from "./CalendarTabs";

describe("CalendarTabs", () => {
    const onCalendarClick = () => ({} as any);

    it("should render with the 'calendar' tab active when the 'calendar' page is active", () => {
        const component = renderer.create(
            <CalendarTabs
                activePage="calendar"
                onCalendarClick={onCalendarClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should handle click on 'calendar' tab", () => {
        const component = renderer.create(
            <CalendarTabs
                activePage="calendar"
                onCalendarClick={onCalendarClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("is-active") !== -1
        ).children[0] as renderer.ReactTestInstance;
        expect(button.props.onClick).toBe(onCalendarClick);
    });
});
