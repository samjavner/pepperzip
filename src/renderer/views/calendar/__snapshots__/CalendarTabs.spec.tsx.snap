// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarTabs should render with the 'calendar' tab active when the 'calendar' page is active 1`] = `
<div
  className="tabs is-centered"
  style={
    Object {
      "marginBottom": "0px",
    }
  }
>
  <ul>
    <li
      className="is-active"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="icon is-small"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-calendar fa-w-14 "
            data-icon="calendar"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 192h424c6.6 0 12 5.4 12 12v260c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V204c0-6.6 5.4-12 12-12zm436-44v-36c0-26.5-21.5-48-48-48h-48V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H160V12c0-6.6-5.4-12-12-12h-40c-6.6 0-12 5.4-12 12v52H48C21.5 64 0 85.5 0 112v36c0 6.6 5.4 12 12 12h424c6.6 0 12-5.4 12-12z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        Calendar
      </a>
    </li>
  </ul>
</div>
`;
